PyVALION/library.py

```diff
@@ -687,6 +691,1343 @@ def find_residuals(model, G, obs_data, obs_info, units):
+
+    # Jason-2 and Jason-3 catalog URLs (THREDDS):
+    catalog_urls = [
+        "https://www.ncei.noaa.gov/thredds-ocean/catalog/jason2/gdr/gdr/"
+        "catalog.xml",
+        "https://www.ncei.noaa.gov/thredds-ocean/catalog/jason3/gdr/gdr/"
+        "catalog.xml",
+        "https://www.ncei.noaa.gov/thredds-ocean/catalog/jason3/gdr/gdr/gdr/"
+        "catalog.xml"
+    ]
+
+    print("Counting total cycle catalogs across all Jason-2 and Jason-3 "
+          "sources...")
+    total_cycles = sum(count_cycle_subcatalogs(url) for url in
+                       catalog_urls)
+    print(f"Total of {total_cycles} cycle catalogs will be scanned.")
+
+    all_nc_urls = []
+
+    # Use progress bar for combined processing
+    with tqdm(total=total_cycles, desc="Processing cycles") as progress:
+        for url in catalog_urls:
+            all_nc_urls.extend(list_nc_files(url, progress=progress))
+
+    # Deduplicate by filename
+    seen = {}
+    for url in all_nc_urls:
+        fname = os.path.basename(url)
+        if (fname not in seen) and "JA2_GPN_2Pf" not in fname:
+            seen[fname] = url  # Save only first occurrence
+
+    # Sort by cycle number
+    sorted_urls = sorted(seen.values(), key=extract_first_date)
+
+    with open(output_file, "w") as f:
+        for url in sorted_urls:
+            f.write(url + "\n")
+
+    print(f"{len(sorted_urls)} unique .nc URLs saved to '{output_file}'")
+
+
+# -----------------------------------------------------------------------------
+# -----------------------------------------------------------------------------
+def update_manifest(output_file):
+    """Update a manifest file by appending new .nc files from newer cycles.
+
+    Scans 'cycleXXX' folders with cycle numbers equal to or higher than the
+    last entry and appends only files with timestamps newer than the most
+    recent entry in the manifest.
+
+    Parameters
+    ----------
+    output_file : str
+        Path to the manifest file to update.
+    """
+    # Print message to user
+    print(f"File '{output_file}' already exists. Checking for updates.")
+
+    # Load existing URLs
+    with open(output_file, "r") as f:
+        existing_urls = [line.strip() for line in f if line.strip()]
+
+    if not existing_urls:
+        print("Jason file manifest is empty. Creating new manifest")
+        create_manifest(output_file)
+        return
+
+    # Get cycle number and timestamp from last URL in manifest
+    last_url = existing_urls[-1]
+    last_cycle = extract_cycle_number(last_url)
+    last_timestamp = extract_first_date(last_url)
+
+    # Check that the final file is not a Jason-2 entry
+    if "JA2" in last_url:
+        print("WARNING: Last listed file is a Jason-2 entry. Rebuilding "
+              "manifest...")
+        create_manifest(output_file)
+        return
+
+    # Hardcoded threshold datetime fir Jason-2 file end
+    # (do not update Jason-2 manifest listings)
+    threshold_str = "20191001_065045"
```

Comment:
I'd recommend declaring threshold variables like this at the top of the file (outside all the functions), so that they are easy to find if/when the remote repository changes things.  If you decide to do this, probably rename it `jason2_threshold` or something similar.